### PR TITLE
Abstract python sampler

### PR DESCRIPTION
**Pull request by gcarleo**
This is PR defines an abstract python class (interface) for samplers (see discussion here #306 ). 

The abstract class is relatively simple and is found in netket/sampler/abstract_sampler.py . Basically it is implemented as a generator, as also suggested by @noamwies . 

There is also a pure-python implementation of samplers provided with this PR. Most notably, all MetropolisHastings sampler (non Parallel Tempering) and exact samplers have pure python versions. 
The backend for the sampler is chosen automatically, depending on the type of the machine to be sampled from.  This is similar to the mechanism implemented by @femtobit for optimizers, but it's hidden in the class implementations. 

A good side effect of this is that now one can use an arbitrary (pure python) MetropolisHastings sampler in case when the machine is not a C++ machine. The speed of the pure python MetropolisHastings is typically comparable to the pure C++ implementation, for machines that are not "too easy" (say, a RBM with at least alpha=4 for L>20 sites), otherwise it's a bit slower.

### CODE REVIEW

File: netket/__init__.py [hunk 1]
```diff
@@ -12,6 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from __future__ import absolute_import
```
Comment (femtobit) [inline netket/__init__.py]: Aren't we ending support for Python 2? This line has no effect in Python 3.

**Comment by gcarleo** [comment]
I was just trying to understand why the import fails on Travis... Any idea?
It works on mac

**Comment by femtobit** [comment]
> I was just trying to understand why the import fails on Travis... Any idea?

I think you need to add the sub-module to `setup.py`, as [it has been done for machine](https://github.com/netket/netket/commit/3d1452bdfe9720e6b14e5b17c1aa7b6a3eed485e), which also has its own subdirectory.

**Comment by gcarleo** [comment]
> I think you need to add the sub-module to `setup.py`, as [it has been done for machine](https://github.com/netket/netket/commit/3d1452bdfe9720e6b14e5b17c1aa7b6a3eed485e), which also has its own subdirectory.

Thank you @femtobit this was very helpful!

**Comment by gcarleo** [comment]
@noamwies is this PR ok for what concerns the python interface of samplers? Do you think we can merge this? Then we can think about the general interface of `machine` and see how to make it interact nicely with FlowKet

**Comment by gcarleo** [comment]
Alright, I would then merge this, so that we can move forward with all the rest, including an abstract interface for machines in python. Anybody is against merging? :)

**Pull request merged**